internal/consensus/state.go

```diff
@@ -1357,7 +1368,11 @@ func (cs *State) defaultDoPrevote(height int64, round int32) {
 	*/
 	if cs.Proposal.POLRound == -1 {
 		if cs.LockedRound == -1 {
-			// TODO(@wbanfield) add check for timely here as well
+			if !cs.proposalIsTimely() {
```

Comment:
Should we do this check only if the node is a validator? We often see full nodes falling behind a few heights, e.g. due to queries, and if they fail the timely check here the only recovery is restart.

Reply:
> if they fail the timely check here the only recovery is restart.

Would you mind elaborating on this a bit? Is the problem that they'll get stuck because the proposal will never appear timely to them?

Reply:
I think a non-validator node does not reach this point of the code, this check is done earlier, isn't it?

Reply:
As far as I can tell, a non-validator does get here.
The code does skip validating the proposal if it is not a validator. See: https://github.com/tendermint/tendermint/blob/e065fba752d29447d01507114f58c23a00df5a99/internal/consensus/state.go#L1154

But, it looks to still collect votes if it's not a validator.

As for the check here, it shouldn't be a problem. This code just defines whether or not the non-validator should prevote for the block. However, the non-validator would never have prevoted for the block (since it's not a validator). It will still commit the block later once all of the precommits from validators _are_ collected. This check should therefore be safe for non-validators.

Reply:
Ah, ok, any node validates a block, true. The `signAddVote` method does nothing for non-validators.

Maybe a valid question here is whether to validate a block (external validation) should occur after the timely validation of a proposal, which should indeed be restricted to validator nodes.

Reply:
Ah, got confused earlier when trying to inspect the code from a full node perspective. The check is indeed safe for non-validators. Looks like the full `defaultDoPrevote` can be skipped for full nodes since the result of all checks here dictate how the node prevotes (`cs.signAddVote`) which is a noop for non-validators.
Also agree with @carson, `ValidateBlock` which is the expensive check should happen after timely.